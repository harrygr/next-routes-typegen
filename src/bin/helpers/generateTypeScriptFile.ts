import walkSync from "walk-sync";

import getFileContent from "./getFileContent";
import getRoutes from "./getRoutes";

const ignorePagesRoutes = ["_app.tsx", "_document.tsx"];
const shouldIncludePageEntry = (route: string) =>
  route.endsWith(".tsx") && !ignorePagesRoutes.includes(route);
const shouldIncludeApiRouteEntry = (endpoint: string) =>
  endpoint.endsWith(".ts");

const getApiRouteFiles = (pagesDir: string) => {
  try {
    return walkSync(`${pagesDir}/api`, {
      directories: false,
    });
  } catch (err) {
    // api routes are not required
    if ((err as any).code === "ENOENT") {
      return [];
    }
    throw err;
  }
};

const generateTypeScriptFile = (pagesDir: string) => {
  const pagesFiles = walkSync(pagesDir, {
    directories: false,
    ignore: ["api"],
  });
  const apiRouteFiles = getApiRouteFiles(pagesDir);
  const relevantPages = pagesFiles.filter(shouldIncludePageEntry);
  const pages = getRoutes(relevantPages.map((page) => `/${page}`));
  const relavantApiRoutes = apiRouteFiles.filter(shouldIncludeApiRouteEntry);
  const apiRoutes = getRoutes(
    relavantApiRoutes.map((route) => `/api/${route}`)
  );

  const fileContent = getFileContent({ pages, apiRoutes });

  return fileContent;
};

export default generateTypeScriptFile;
