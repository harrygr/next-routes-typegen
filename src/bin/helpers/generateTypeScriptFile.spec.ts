import { describe, expect, it } from "vitest";
import generateTypeScriptFile from "./generateTypeScriptFile";

describe("generateTypeScriptFile", () => {
  it("generates types for pages directory with a single page", () => {
    const types = generateTypeScriptFile(`${__dirname}/mocks/pages-no-api`);

    expect(types).toMatchInlineSnapshot(`
      "// Do not modify this file. It is auto-generated.
      //
      // To regenerate it run the \`generate-route-types\` command.

      declare module 'next-routes-typegen' {
        type Query = Record<string, any>;
        
        export type TypeSafePage = 
          |  \\"/\\"
          |  { route: \\"/\\", query?: Query, path?: undefined }

        export type TypeSafeApiRoute = TypeSafePage

        export const getRoute: (typeSafeUrl: TypeSafePage | TypeSafeApiRoute) => string;
        export const getPathname: (typeSafeUrl: TypeSafePage | TypeSafeApiRoute) => string;
      }

      "
    `);
  });

  it("generates types for pages directory with pages and api routes", () => {
    const types = generateTypeScriptFile(`${__dirname}/mocks/pages`);

    expect(types).toMatchSnapshot();
  });

  it("generates types for an empty pages directory", () => {
    const types = generateTypeScriptFile(`${__dirname}/mocks/empty-pages`);

    expect(types).toMatchInlineSnapshot(`
      "// Do not modify this file. It is auto-generated.
      //
      // To regenerate it run the \`generate-route-types\` command.

      declare module 'next-routes-typegen' {
        type Query = Record<string, any>;
        
        export type TypeSafePage = never

        export type TypeSafeApiRoute = TypeSafePage

        export const getRoute: (typeSafeUrl: TypeSafePage | TypeSafeApiRoute) => string;
        export const getPathname: (typeSafeUrl: TypeSafePage | TypeSafeApiRoute) => string;
      }

      "
    `);
  });
});
