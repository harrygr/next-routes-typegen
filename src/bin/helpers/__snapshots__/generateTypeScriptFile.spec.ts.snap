// Vitest Snapshot v1

exports[`generateTypeScriptFile > generates types for pages directory with pages and api routes 1`] = `
"// Do not modify this file. It is auto-generated.
//
// To regenerate it run the \`generate-route-types\` command.

declare module 'next-routes-typegen' {
  type Query = Record<string, any>;
  
  export type TypeSafePage = 
    |  \\"/404\\"
    |  { route: \\"/404\\", query?: Query, path?: undefined }
    |  { route: \\"/catch-all\\", path: string, query?: Query }
    |  \\"/\\"
    |  { route: \\"/\\", query?: Query, path?: undefined }
    |  { route: \\"/nested-catch-all/[dynamic]/slugs\\", path: string, params: { \\"dynamic\\": string | number }, query?: Query }
    |  \\"/optional-catch-all\\"
    |  { route: \\"/optional-catch-all\\", path?: string, query?: Query }
    |  { route: \\"/route-with-symbols/[value-with-hyphens]\\", params: { \\"value-with-hyphens\\": string | number }, query?: Query, path?: undefined }
    |  { route: \\"/users/[userId]\\", params: { \\"userId\\": string | number }, query?: Query, path?: undefined }
    |  \\"/users\\"
    |  { route: \\"/users\\", query?: Query, path?: undefined }

  export type TypeSafeApiRoute = 
    |  { route: \\"/api/[authId]\\", params: { \\"authId\\": string | number }, query?: Query, path?: undefined }
    |  { route: \\"/api/catch-all\\", path: string, query?: Query }
    |  \\"/api/multiple-periods/site.webmanifest\\"
    |  { route: \\"/api/multiple-periods/site.webmanifest\\", query?: Query, path?: undefined }
    |  \\"/api/optional-catch-all\\"
    |  { route: \\"/api/optional-catch-all\\", path?: string, query?: Query }
    |  { route: \\"/api/users/[userId]\\", params: { \\"userId\\": string | number }, query?: Query, path?: undefined }
    |  \\"/api/users\\"
    |  { route: \\"/api/users\\", query?: Query, path?: undefined }

  export const getRoute: (typeSafeUrl: TypeSafePage | TypeSafeApiRoute) => string;
  export const getPathname: (typeSafeUrl: TypeSafePage | TypeSafeApiRoute) => string;
}

"
`;
